thread/thread11.h

```diff
@@ -24,160 +25,49 @@ limitations under the License.
+    std::__invoke(std::move(std::get<I0>(t)), std::move(std::get<I>(t))...);
+}
 
-    template <typename CLASS, typename F, typename... ARGUMENTS>
-    inline std::enable_if_t<std::is_member_function_pointer<F>::value, thread*>
-    thread_create11(F f, CLASS* obj, ARGUMENTS&&... args) {
-        return thread_create11<CLASS, F, ARGUMENTS...>(
-            DEFAULT_STACK_SIZE, f, obj, std::forward<ARGUMENTS>(args)...);
-    }
+template <class Tuple>
+void* __thread_proxy(void* vp) {
+    // Tuple = std::tuple<Func, Args...>
+    std::unique_ptr<Tuple> p(static_cast<Tuple*>(vp));
 
-    template <typename FUNCTOR, typename... ARGUMENTS>
-    inline void __functor_call_helper(
-        typename std::decay<FUNCTOR>::type f,
-        typename std::decay<ARGUMENTS>::type... args) {
-        f(std::forward<ARGUMENTS>(args)...);
-    }
+    __thread_execute(*p, std::make_index_sequence<std::tuple_size<Tuple>::value>{});
+    return nullptr;
+}
 
-    // NOTICE: The arguments to the thread function are moved or copied by
-    // value. If a reference argument needs to be passed to the thread function,
-    // it has to be wrapped (e.g., with std::ref or std::cref).
-    template <typename T, typename... Args>
-    struct is_functor
-        : std::conditional<
-              std::is_class<typename std::decay<T>::type>::value &&
-                  std::is_constructible<typename std::function<void(Args...)>,
-                                        typename std::reference_wrapper<
-                                            typename std::remove_reference<
-                                                T>::type>::type>::value,
-              std::true_type, std::false_type>::type {};
-
-    template <typename FUNCTOR, typename... ARGUMENTS>
-    inline typename std::enable_if<is_functor<FUNCTOR, ARGUMENTS...>::value,
-                                   thread*>::type
-    thread_create11(uint64_t stack_size, FUNCTOR&& f, ARGUMENTS&&... args) {
-        // takes `f` as parameter to helper function
-        // thread_create11 will make sure parameters copy is completed
-        return thread_create11(
-            stack_size, &__functor_call_helper<FUNCTOR, ARGUMENTS...>,
-            std::forward<FUNCTOR>(f), std::forward<ARGUMENTS>(args)...);
+// NOTICE: The arguments to the thread function are moved or copied by
+// value. If a reference argument needs to be passed to the thread function,
+// it has to be wrapped (e.g., with std::ref or std::cref).
+template <class F, class... Args, typename = std::enable_if_t<std::__is_invocable<F&&, Args&&...>::value>>
+thread* thread_create11(uint64_t stack_size, F&& f, Args&&... args) {
+    using Gp = std::tuple<typename std::decay<F>::type, typename std::decay<Args>::type...>;
+    std::unique_ptr<Gp> p(new Gp(std::forward<F>(f), std::forward<Args>(args)...));
+    auto thread = thread_create(&__thread_proxy<Gp>, p.get(), stack_size);
```

Comment:
unique_ptr `p` will be destruct once `thread_create11` returned.
But at this time, `thread` did not launched as STL `std::thread` since no certainlly yield to thread.

when new thread launched and copying context, the context struct already released, leads to segment fault.

Reply:
> 

unique_ptr `p` will release

Reply:
https://en.cppreference.com/w/cpp/memory/unique_ptr/release

Reply:
`release` will reset pointer without destruct. It will not cause segment fault. the comment is my fault.

On heap context needs extra allocation action may affect performance.
I will test it and reconsidering the PR.

Thank you.

Reply:
但这样少了两次协程切换不是吗？标注库不也是存储在堆上的吗？这个无伤大雅吧

Reply:
本身我认为问题不大，性能上可能会有益处。但考虑使用切换代替堆上申请是原本设计中的一些考虑，因此需要一些权衡。
还有一处担心在于，先前的实现中除functor类型的F外，其他F接受参数类型都是由函数原型的参数表指定的，而现在由thread_create11的参数类型决定。这也是为何后续大量测试改为显式标注std::ref的原因。
我将后续进行一些性能及现有代码兼容性上的检查和测试，确认后考虑合并，还请稍等一小段时间。
同时我也会考虑寻找是否存在除直接使用internal implemention之外的解决方法。